dali/kernels/imgproc/warp_cpu.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_CPU_H_
+#define DALI_KERNELS_IMGPROC_WARP_CPU_H_
+
+#include "dali/core/common.h"
+#include "dali/core/geom/vec.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+#include "dali/kernels/imgproc/sampler.h"
+#include "dali/kernels/imgproc/warp/map_coords.h"
+
+namespace dali {
+namespace kernels {
+
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+          typename _BorderType>
+class WarpCPU {
+ public:
+  static constexpr int spatial_ndim = ndim;
+  static constexpr int tensor_ndim = ndim + 1;
+  static constexpr int channel_dim = ndim;
+
+  static_assert(spatial_ndim == 2, "Not implemented for spatial_ndim != 2");
```

Comment:
```suggestion
  static_assert(spatial_ndim == 2, "This kernel's implementation assumes HWC layout, therefore spatial_ndim must be equal to 2");
```
It's not that this kernel is not implemented for `spatial_ndim!=2`, but it **is implemented for `spatial_ndim==2`**. Logically there's no difference, but semantically there is...

Reply:
1. This is _temporary_. It it wasn't, the ndim parameter would even be here.
2. Your proposed message is misleading. The limitation on number of dimensions is not a consequence of layout constraint, but lack of "Surface3D" and a corresponding sampler.
I can rephrase it to "This kernel is implemented only for 2D data." - however, the phrase "Not implemented" usually carries an additional message which reads "not implemented *yet*". Stating that the implementation assumes something doesn't suggest that it was ever meant to support something else.

Reply:
I'd see no point of making the message suggest, that at some point in the future (not really clearly defined) we will add some functionality, because:
1. It's not set up, when we do it
2. When we decide "ok, we may do it" this can be no longer needed

My message is just a suggestion, where I wanted to say, that this kernel is actually implemented only regarding to 3-dimension (including channels) image, which serves as a remark.

Reply:
I don't want to mix layout with number of spatial dimensions. It's just confusing.